src/main/java/it/pagopa/pn/deliverypush/action/notificationview/NotificationViewedRequestHandler.java

```diff
@@ -111,4 +118,63 @@ private PnAuditLogEvent generateAuditLog(String iun, Integer recIndex, RaddInfo
                 .build();
     }
 
+    public Mono<Void> handleNotificationRaddRetrieved(String iun, RequestNotificationViewedDto requestNotificationViewedDto) {
+        return Mono.fromCallable(() -> notificationService.getNotificationByIun(iun))
+                .flatMap( notification -> {
+                    int recIndex = notificationUtils.getRecipientIndexFromInternalId(notification, requestNotificationViewedDto.getRecipientInternalId());
+
+                    RaddInfo raddInfo = RaddInfo.builder()
+                            .type(requestNotificationViewedDto.getRaddType())
+                            .transactionId(requestNotificationViewedDto.getRaddBusinessTransactionId())
+                            .build();
+                    return handleViewNotificationRaddRetrieved(iun, recIndex, requestNotificationViewedDto.getRecipientInternalId(), raddInfo, requestNotificationViewedDto.getRaddBusinessTransactionDate());
+                });
+    }
+
+    private Mono<Void> handleViewNotificationRaddRetrieved(String iun, Integer recIndex, String recipientInternalId, RaddInfo raddInfo, Instant eventTimestamp) {
+        return Mono.fromCallable(() -> timelineUtils.checkIsNotificationCancellationRequested(iun))
+                .flatMap(isNotificationCancelled -> {
+                    if (Boolean.TRUE.equals(isNotificationCancelled)) {
+                        log.warn("For this notification a cancellation has been requested - iun={} id={}", iun, recIndex);
+                        return Mono.empty();
+                    }
+
+                    return Mono.just(timelineUtils.checkIsNotificationViewed(iun, recIndex));
+                })
+                .flatMap(isNotificationAlreadyViewed -> {
+                    //I processi collegati alla visualizzazione di una notifica vengono effettuati solo la prima volta che la stessa viene visualizzata
+                    if (Boolean.FALSE.equals(isNotificationAlreadyViewed)) {
+
+                        PnAuditLogEvent logEvent = generateRaddRetrievedAuditLog(iun, recipientInternalId);
+                        logEvent.log();
```

Comment:
Io porterei l'auditlog ad inizio del metodo.
Poi, nel caso di errori loggare un logError, mentre nel caso in cui qualche check non sia valido (esempio quello della notifica annullata), metterei un logWarning.
Eventualmente, loggare logWarning anche nel caso di addTimelineElement che ritorna "false"